### ISSUE TITLE
Provide option via config to make `fetch` shallow instead of unshallow

### ISSUE THREAD

**Issue by Aghassi** [semantic-release/semantic-release#934]
## New feature motivation

We have a repo that has a history of 1.5GB+. Unshallow causes our CI times to be 3+ minutes.

## New feature description

Would like the ability to specify a `shallow: true` in the `release.config.js` file that would cause this command to not run `unshallow` by default.

## New feature implementation

In your `release.config.js`, accept a config setting named `shallow`. By default this can be set to `false`, but should have the ability to be set to `true`. If set to `true`, change the `fetch` command to do a `shallow` lookup instead of `unshallow`. It would also be nice to be able to specify depth (though I don't know how necessary this is up front. So your config could look like:

```json
        "prepare": [
            "@semantic-release/npm",
            "@semantic-release/git"
        ],
        "shallow": true,
        "depth": 10
```

## References

Usage
https://github.com/semantic-release/semantic-release/blob/caribou/index.js#L78

Implementation
https://github.com/semantic-release/semantic-release/blob/caribou/lib/git.js#L65

**Comment by pvdlg** [comment]
The problem is that if we limit the depth we might not be able to properly determine the last version released.
For example is configure a depth of 10 and you push 10 commits since the last release semantic-release will not fing the commit used for that last release and will consider there is no previous one, so it will release `1.0.0`.

Unshallowing will also be very important for #563 as we would need to check the tags on multiple branch, some of them potentially really old. For example if your last release is `3.2.1` and for some reasons you need to deliver a fix for a client that can't upgrade to `3.x.x`. So you create a new branch from the last `2.x.x` tag, let's say `2.3.4`, apply your fix there and release `2.3.5`.
If `2.3.4` was release 1 year ago, you probably have 100s of new commit in your repo since the commit tagged as `2.3.4`. semantic-release would need to unshallow with a huge depth value to be sure to have it included.

In addition, and without going into too much technical details here, in #563 we have to do a lot verification for consistency across your branches (i.e. if you have a `2.x.x` branch and a `master` you don't want to release a version from `2.x.x` that is higher than the last version release from `master`).
Doing those checks require to have a long history of commits in the local repo, as we need to have the last version tagged commit on each branch.

We used to unshallow with a depth of 50 before (maybe a year ago) and we changed to a full unshallow as it was creating to many issues for a lot of users. Those issues were pretty bad as they would end up releasing a wrong version.

Even if what you propose is only an option, I think it opens the door to many bigger problems and there is probably better solutions.

Maybe there is a way to configure your CI to cache the local repo, so on new builds instead of cloning the repo it would get it from the cache and fetch/pull the new commits.

**Comment by pvdlg** [comment]
Also 1.5G is pretty extreme, and will probably encounter other issues (like limitation on the Git Hosting platform) so you might want to look at why it's so big maybe see if you can reduce the size.
See https://confluence.atlassian.com/bitbucket/reduce-repository-size-321848262.html for some suggestions.

**Comment by Aghassi** [comment]
@pvdlg Don't disagree. Just looking at ways to avoid rewriting history to trim the size of the repo. Would mean all our contributors would need to force reset their masters etc and cherry pick their features. It's just a bit of a massive headache which is why I posed this question. Is there a way semantic-release can leave a "trail" of when the last time it was run, and use that as the shallow clone depth? So parse the git log backwards till you find the last release?. I understand there may be issues on multiple branches, I'm just trying to see if there is a flexible medium (I am by means no expert in the area).

**Comment by pvdlg** [comment]
> Is there a way semantic-release can leave a "trail" of when the last time it was run, and use that as the shallow clone depth?

The only state shared between multiple semantic-release runs are the Git tags. In any case there is no way store that information when semantic-release runs, because the info we need comes from the remote repo (e.g. new commits, new branches, new tags etc...)

> So parse the git log backwards till you find the last release?

That is not possible because depending on the merge strategy you use the commits will not be ordered by their merge time.
If you create a PR with a commit from Sept 23 and another one with a commit from Sept 24, merge that second PR which release `1.0.0` and then merge the first PR which release `1.0.1`, then if you list the tags chronologically you will get `1.0.1` happening before `1.0.0`. That's because the tags do not have a date, only the commit do.

**Comment by Aghassi** [comment]
I'm surprised. It seems tagging supports dates according to this https://git-scm.com/docs/git-tag#_on_backdating_tags to some extent. Also, tags are associated with commits right? Could we not infer the date from the time of the commit?

Unfortunately, dates make everything difficult. 😞 

Pardon my questions, I still have a lot to learn in this area and wanted to best understand the problem before fully conceding this isn't possible 😅

**Comment by pvdlg** [comment]
Yes the tag have a date, but fetch with `--depth` is based on the commit, not the tags. Because the problem explain the commits might not be ordered the same as the tag.

So if you'd be fetching commit 10 by 10, and check if you find a tag in the newly fetched ones, there is no guarantee the first tag you'll find is the last one created.

Also, even if that was possible, once #563 is implemented, suffice. In #563 we check on each branch configured to release if there is a tag that was created from another branch but not the current one.

For example if you have a `master` and a `next` branch we go through all the tags on `master` to see if one was created on `next` and not on `master` (tags created on `master` are named `vx.y.z`, the one created on `next` are named `vx.y.z@next`). If we find one of those we make the corresponding version available on the dist-tag associated with master.

So if you release `1.1.0` from `next` (on the dist-tag `@next`) we create the tag `v1.1.0@next`. When you merge `next` in to `master`, we go through all the tags on `master` and we find that `v1.1.0@next` is in `master` history (because of the merge) so we search for `v1.1.0` in `master` history. We don't find it and therefore conclude that `1.1.0` was release from `next` but not from `master`, so we add that release to `@latest` and create the tag `v1.1.0` on the commit where `v1.1.0@next` is.

As you can see, in that process, we need all the commits in the repo in order to:
- finding all the `vx.y.x@next` tags in `master` history
- test if `v1.1.0` exist on `master`

Basically we need to keep a state in between run, with the info of what was released from where. Using tags is much more robust than any other solution because it's embedded in Git. Using a DB or a file, would create a lot more complexity and potential for error (like if you loose that state, you can't reliably release anything new with semantic-release).

**Comment by pvdlg** [comment]
@Aghassi did you find an alternative solution that wouldn't cause the problem mentioned?
On my side I've been thinking about that for a couple week and I don't see any other solution that write the state outside of the git repository, which is not a good solution, so we wouldn't implement it.
